datafusion-examples/examples/parquet_encryption_with_kms.rs

```diff
@@ -0,0 +1,216 @@
+    CryptoFactory, DecryptionConfiguration, EncryptionConfiguration,
+};
+use parquet_key_management::kms::KmsConnectionConfig;
+use parquet_key_management::test_kms::TestKmsClientFactory;
+use std::collections::HashMap;
+use std::fmt::Formatter;
+use std::sync::Arc;
+use tempfile::TempDir;
+
+/// This example demonstrates reading and writing Parquet files that
+/// are encrypted using Parquet Modular Encryption, and uses the
+/// parquet-key-management crate to integrate with a Key Management Server (KMS).
+
+const ENCRYPTION_FACTORY_ID: &'static str = "example.inmem_kms_encryption";
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let ctx = SessionContext::new();
+
+    // Register an `EncryptionFactory` implementation to be used for Parquet encryption
+    // in the session context.
+    // This example uses an in-memory test KMS from the `parquet_key_management` crate with
+    // a custom `KmsEncryptionFactory` wrapper type to integrate with DataFusion.
+    // `EncryptionFactory` instances are registered with a name to identify them so
+    // they can be later referenced in configuration options, and it's possible to register
+    // multiple different factories to handle different ways of encrypting Parquet.
+    // In future it could be possible to have built-in implementations in DataFusion.
+    let crypto_factory = CryptoFactory::new(TestKmsClientFactory::with_default_keys());
+    let encryption_factory = KmsEncryptionFactory { crypto_factory };
+    ctx.register_parquet_encryption_factory(
+        ENCRYPTION_FACTORY_ID,
+        Arc::new(encryption_factory),
+    );
+
+    let tmpdir = TempDir::new()?;
+    write_encrypted(&ctx, &tmpdir).await?;
+    read_encrypted(&ctx, &tmpdir).await?;
+    Ok(())
+}
+
+/// Write an encrypted Parquet file
+async fn write_encrypted(ctx: &SessionContext, tmpdir: &TempDir) -> Result<()> {
+    let a: ArrayRef = Arc::new(StringArray::from(vec!["a", "b", "c", "d"]));
+    let b: ArrayRef = Arc::new(Int32Array::from(vec![1, 10, 10, 100]));
+    let batch = RecordBatch::try_from_iter(vec![("a", a), ("b", b)])?;
+
+    ctx.register_batch("test_data", batch)?;
+    let df = ctx.table("test_data").await?;
+
+    let mut parquet_options = TableParquetOptions::new();
+    // We specify that we want to use Parquet encryption by setting the identifier of the
+    // encryption factory to use.
+    //parquet_options.encryption.factory_id = ENCRYPTION_FACTORY_ID.to_owned();
+    parquet_options.global.encryption_factory_id = Some(ENCRYPTION_FACTORY_ID.to_owned());
+    // Our encryption factory requires specifying the master key identifier to
+    // use for encryption. To support arbitrary configuration options for different encryption factories,
+    // DataFusion could use a HashMap<String, String> field for encryption options.
+    //let encryption_config = EncryptionConfig::new("kf".to_owned());
+    //parquet_options.encryption.factory_options = encryption_config.to_config_map();
```

Comment:
As a user, this is a bit hard to follow. Maybe make this part of the example?

Reply:
This is commented out still as I hadn't implemented any config handling and had just hard coded the master key identifier in the `KmsEncryptionFactory` logic for now, but it should end up in the example once it's working properly.